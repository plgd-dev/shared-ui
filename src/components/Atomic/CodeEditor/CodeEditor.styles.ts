import { css } from '@emotion/react'
import { ThemeType, get } from '../_theme'

export const editor = (theme: ThemeType) => css`
    border: 1px solid ${get(theme, `CodeEditor.borderColor`)};
    border-radius: 8px;
    overflow: hidden;
    position: relative;
    z-index: 2;

    .cm-editor {
        background: transparent;
        transition: all 0.3s;
    }

    .cm-gutters {
        background: ${get(theme, `CodeEditor.gutters.background`)};
        border: 0;
    }

    .cm-lineNumbers .cm-gutterElement {
        color: ${get(theme, `CodeEditor.color`)};
        text-align: center;
        padding: 0 16px 0 16px;
    }

    .cm-gutterElement {
        color: ${get(theme, `CodeEditor.gutters.color`)}!important;
    }

    .cm-content {
        padding: 16px 0;
    }

    .cm-line {
        padding: 0 16px;
        line-height: 1.6;
    }
`

export const wrapper = css`
    position: relative;
`

export const background = (theme: ThemeType) => css`
    .cm-editor {
        background: ${get(theme, `CodeEditor.background`)};
    }
`

export const placeholder = (theme: ThemeType) => css`
    position: absolute;
    border-radius: 8px;
    top: 0;
    left: 0;
    bottom: 0;
    right: 0;
    background-color: ${get(theme, `CodeEditor.background`)};
    width: 100%;
    height: 100%;
    z-index: 1;
    display: flex;
    align-items: center;
    justify-content: center;
`

export const flex = css`
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    transition: all 0.3s;
    opacity: 1;
`

export const noPlaceholder = css`
    opacity: 0;
`

export const placeholderText = (theme: ThemeType) => css`
    color: ${get(theme, `CodeEditor.placeholder.color`)};
    font-weight: 400;
    font-size: 14px;
    line-height: 22px;
`
