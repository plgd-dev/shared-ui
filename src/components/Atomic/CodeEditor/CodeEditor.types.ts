export type Props = {
    disabled?: boolean
    height?: string | number
    onChange?: (value: string) => void
    placeholderText?: string
    value: string
}

export const defaultProps: Partial<Props> = {
    height: '500px',
}
