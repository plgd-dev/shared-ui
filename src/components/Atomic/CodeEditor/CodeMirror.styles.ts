import { css } from '@emotion/react'
import { ThemeType, get } from '../_theme'

export const editor = (theme: ThemeType) => css`
    border: 1px solid ${get(theme, `Editor.borderColor`)};
    border-radius: 8px;
    overflow: hidden;

    .cm-gutters {
        background: #e6e9ed;
        border: 0;
    }

    .cm-lineNumbers .cm-gutterElement {
        color: #757676;
        text-align: center;
        padding: 0 16px 0 16px;
    }

    .cm-content {
        padding: 16px 0;
    }

    .cm-line {
        padding: 0 16px;
        line-height: 1.6;
    }
`
