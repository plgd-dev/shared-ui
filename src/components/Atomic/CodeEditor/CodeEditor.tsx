import { FC, useCallback, useMemo } from 'react'
import CodeMirror from '@uiw/react-codemirror'
import { loadLanguage } from '@uiw/codemirror-extensions-langs'
import isFunction from 'lodash/isFunction'
import { tags as t } from '@lezer/highlight'
import { createTheme } from '@uiw/codemirror-themes'
import { useTheme } from '@emotion/react'

import { Props, defaultProps } from './CodeEditor.types'
import * as styles from './CodeEditor.styles'
import { getSizeInPx } from '../_utils/commonStyles'
import { ThemeType } from '../_theme'
import { convertSize, IconFileUpload } from '../Icon'

const CodeEditor: FC<Props> = (props) => {
    const { disabled, height, onChange, placeholderText, value } = { ...defaultProps, ...props }
    const globalTheme: ThemeType = useTheme()

    const handleChange = useCallback(
        (v: string) => {
            isFunction(onChange) && onChange(v)
        },
        [onChange]
    )

    const extensions = useMemo(() => {
        const ext = []

        if (isFunction(loadLanguage)) {
            ext.push(loadLanguage('markdown') as any)
        }

        return ext
    }, [])

    const theme = useMemo(
        () =>
            createTheme({
                theme: 'light',
                settings: {
                    background: globalTheme?.colors?.CodeEditor.background,
                    lineHighlight: globalTheme?.colors?.CodeEditor.lineHighlight,
                    selection: globalTheme?.colors?.CodeEditor.selection,
                },
                styles: [
                    { tag: t.comment, color: '#787b8099' },
                    { tag: t.variableName, color: '#0080ff' },
                    { tag: [t.string, t.special(t.brace)], color: '#5c6166' },
                    { tag: t.number, color: '#5c6166' },
                    { tag: t.bool, color: '#5c6166' },
                    { tag: t.null, color: '#5c6166' },
                    { tag: t.keyword, color: '#5c6166' },
                    { tag: t.operator, color: '#5c6166' },
                    { tag: t.className, color: '#5c6166' },
                    { tag: t.definition(t.typeName), color: '#5c6166' },
                    { tag: t.typeName, color: '#5c6166' },
                    { tag: t.angleBracket, color: '#5c6166' },
                    { tag: t.tagName, color: '#5c6166' },
                    { tag: t.attributeName, color: '#5c6166' },
                ],
            }),
        [globalTheme?.colors?.CodeEditor.background, globalTheme?.colors?.CodeEditor.lineHighlight, globalTheme?.colors?.CodeEditor.selection]
    )

    const showPlaceholder = useMemo(() => value === '', [value])

    return (
        <div css={styles.wrapper}>
            <CodeMirror
                basicSetup={{
                    foldGutter: false,
                    drawSelection: true,
                }}
                css={[styles.editor, !showPlaceholder && styles.background]}
                editable={disabled === true ? false : undefined}
                extensions={extensions}
                height={getSizeInPx(height!)}
                onChange={(v) => handleChange(v)}
                theme={theme}
                value={value}
            />
            {placeholderText && (
                <div css={styles.placeholder}>
                    <div css={[styles.flex, !showPlaceholder && styles.noPlaceholder]}>
                        <IconFileUpload {...convertSize(50)} />
                        <p css={styles.placeholderText}>{placeholderText}</p>
                    </div>
                </div>
            )}
        </div>
    )
}

CodeEditor.displayName = 'CodeEditor'
CodeEditor.defaultProps = defaultProps

export default CodeEditor
