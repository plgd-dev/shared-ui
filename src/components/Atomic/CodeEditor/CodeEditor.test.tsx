import { render } from '../../../../test/jest-wrapper'
import CodeEditor from './CodeEditor'

describe('<CodeEditor>', () => {
    it('renders without crashing', () => {
        const { container, asFragment } = render(<CodeEditor value='' />)
        expect(container).toBeInTheDocument()

        expect(asFragment()).toMatchSnapshot()
    })
})
