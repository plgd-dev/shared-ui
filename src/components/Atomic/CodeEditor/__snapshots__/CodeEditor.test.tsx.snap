// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CodeEditor> renders without crashing 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
}

.emotion-1 {
  border: 1px solid;
  border-radius: 8px;
  overflow: hidden;
  position: relative;
  z-index: 2;
}

.emotion-1 .cm-editor {
  background: transparent;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
}

.emotion-1 .cm-gutters {
  border: 0;
}

.emotion-1 .cm-lineNumbers .cm-gutterElement {
  text-align: center;
  padding: 0 16px 0 16px;
}

.emotion-1 .cm-gutterElement {
  color: !important;
}

.emotion-1 .cm-content {
  padding: 16px 0;
}

.emotion-1 .cm-line {
  padding: 0 16px;
  line-height: 1.6;
}

.emotion-2 {
  position: absolute;
  border-radius: 8px;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  width: 100%;
  height: 100%;
  z-index: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  opacity: 1;
  opacity: 0;
}

.emotion-4 {
  font-weight: 400;
  font-size: 14px;
  line-height: 22px;
}

<div
    class="emotion-0"
  >
    <div
      class="cm-theme emotion-1"
    >
      <div
        class="cm-editor ͼ1 ͼ2 ͼ4 ͼ1n ͼ16"
      >
        <div
          aria-live="polite"
          class="cm-announced"
        />
        <div
          class="cm-scroller"
          tabindex="-1"
        >
          <div
            aria-hidden="true"
            class="cm-gutters"
            style="min-height: 14px; position: sticky;"
          >
            <div
              class="cm-gutter cm-lineNumbers"
            >
              <div
                class="cm-gutterElement"
                style="height: 0px; visibility: hidden; pointer-events: none;"
              >
                9
              </div>
              <div
                class="cm-gutterElement cm-activeLineGutter"
                style="height: 14px;"
              >
                1
              </div>
            </div>
          </div>
          <div
            aria-autocomplete="list"
            aria-multiline="true"
            autocapitalize="off"
            autocorrect="off"
            class="cm-content"
            contenteditable="true"
            data-language="markdown"
            role="textbox"
            spellcheck="false"
            style="tab-size: 4;"
            translate="no"
          >
            <div
              class="cm-activeLine cm-line"
            >
              <br />
            </div>
          </div>
          <div
            aria-hidden="true"
            class="cm-layer cm-layer-above cm-cursorLayer"
            style="z-index: 150; animation-duration: 1200ms;"
          />
          <div
            aria-hidden="true"
            class="cm-layer cm-selectionLayer"
            style="z-index: -2;"
          />
        </div>
      </div>
    </div>
    <div
      class="emotion-2"
    >
      <div
        class="emotion-3"
      >
        <svg
          fill="none"
          height="50"
          viewBox="0 0 50 61"
          width="50"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M.5 6A5.5 5.5 0 0 1 6 .5h23.046A5.5 5.5 0 0 1 32.9 2.077l3.25 3.192 3.705 3.64a5.5 5.5 0 0 1 1.645 3.923V50a5.5 5.5 0 0 1-5.5 5.5H6A5.5 5.5 0 0 1 .5 50z"
            fill="#F6F7F9"
            stroke="#E6E9ED"
          />
          <mask
            height="56"
            id="icon-file-upload_svg__a"
            maskUnits="userSpaceOnUse"
            style="mask-type: alpha;"
            width="42"
            x="0"
            y="0"
          >
            <path
              d="M.5 6A5.5 5.5 0 0 1 6 .5h23.046A5.5 5.5 0 0 1 32.9 2.077l3.25 3.192 3.705 3.64a5.5 5.5 0 0 1 1.645 3.923V50a5.5 5.5 0 0 1-5.5 5.5H6A5.5 5.5 0 0 1 .5 50z"
              fill="#F6F7F9"
              stroke="#D7D8DA"
            />
          </mask>
          <g
            mask="url(#icon-file-upload_svg__a)"
          >
            <path
              d="M30 0h12v12h-9a3 3 0 0 1-3-3z"
              fill="#E6E9ED"
            />
          </g>
          <circle
            cx="34"
            cy="45"
            fill="#2261AE"
            r="16"
          />
          <path
            d="M29.417 50.833a3.751 3.751 0 0 1-1.245-7.29l-.005-.21a4.166 4.166 0 0 1 7.779-2.078 2.5 2.5 0 0 1 3.721 2.98A3.333 3.333 0 0 1 39 50.832zm2.057-5.651a.625.625 0 0 0 .883.883l1.016-1.016v3.492c0 .346.278.625.625.625a.623.623 0 0 0 .625-.625v-3.492l1.015 1.016a.625.625 0 0 0 .883-.883l-2.083-2.084a.62.62 0 0 0-.883 0l-2.083 2.084z"
            fill="#fff"
          />
        </svg>
        <p
          class="emotion-4"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
